### PR TITLE
[JUJU-4012] Wire up secrets watcher registry

### PR DESCRIPTION
**Pull request by SimonRichardson**
~~Requires https://github.com/juju/juju/pull/15782 to land first. Will rebase after it's landed.~~

Rather than having all the watchers in one file, we move the watchers
into their own dedicated watchers package. This makes is much easier
to reason about and navigate.

The watchers all attempt to first locate the watcher from the watcher
registry, and upon failing that with a not found, will fall back to the
resources. As resources are going to be deprecated we'll hopefully
seeing the usage of that being removed over time.

As an example of this, all secret watchers are moved over to the
watcher registry.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ cd tests
$ ./main.sh secrets_iaas
```

### CODE REVIEW

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 1]
```diff
@@ -893,8 +894,13 @@ func (s *SecretsManagerAPI) WatchObsolete(args params.Entities) (params.StringsW
 		return result, errors.Trace(err)
 	}
 	if changes, ok := <-w.Changes(); ok {
-		result.StringsWatcherId = s.resources.Register(w)
-		result.Changes = changes
+		id, err := s.watcherRegistry.Register(w)
+		if err != nil {
+			result.Error = apiservererrors.ServerError(err)
+		} else {
+			result.StringsWatcherId = id
```
Comment (manadart) [inline apiserver/facades/agent/secretsmanager/secrets.go:901]: Shame there isn't uniformity with the `<Thing>WatchResultTypes` (`StringsWatcherId` here versus `WatcherId` elsewhere). Then we might be able to reduce this boilerplate with generics.

File: state/secrets.go [hunk 2]
```diff
@@ -1999,7 +1999,9 @@ func newObsoleteSecretsWatcher(st modelBackend, owners []string) *obsoleteSecret
 	}
 	w.tomb.Go(func() error {
 		defer w.finish()
-		return w.loop()
+		err := w.loop()
+		fmt.Println("!!!", err)
```
Comment (manadart) [inline state/secrets.go:2003]: Not sure you meant to leave this.

**Review by manadart — COMMENTED** [review]
I know you're still working on this, but I've looked at about half of it so far.

**Comment by SimonRichardson** [comment]
Closing this in favour of https://github.com/juju/juju/pull/15817

File: apiserver/facades/agent/secretsmanager/secrets.go (hunk 1 above)
Comment (SimonRichardson) [reply apiserver/facades/agent/secretsmanager/secrets.go:901]: The whole watcher mess isn't great. Ideally, we should have one watcher per model and emit events on top of the change stream.

**Pull request closed without merging**